### DISCUSSION TITLE
Adding defines as _additional_ preprocessor definitions

### DISCUSSION THREAD

**Discussion by loostrum** [KernelTuner/kernel_tuner#235] [Q&A] [answered]
I have a kernel where I tune a bunch of parameters, but need to insert some static values as well (think problem size etc.) as preprocessor definitions. I thought I'd be able to use the `defines` option for this, but using that means the tunable parameters are no longer inserted. It would be great if there'd be a way to insert both the tunable parameters and some more defines.

Currently I'm using single-value tunable parameters as a workaround, but it clutters the tuning output a bit.

**Comment by stijnh** [comment, marked as answer]
You should be able to use:

```python
defines = dict(
block_size_x=lambda p: p["block_size_x"],
use_smem=lambda p: p["use_smem"],
...
)
```

If I remember correctly, this should also work:

```python
defines = dict(
block_size_x="block_size_x",
use_smem="use_smem",
...
)
```

In which case, you can generate the dict automatically:

```python
defines = {key: key for key in tune_params}
defines["another_variable"] = 1
defines["more_variables"] = 2
```

**Comment by loostrum** [comment]
Thanks for the tip, that should solve my issue!

**Comment by stijnh** [comment]
Great! This should be documented better since currently its a bit hidden away in the API documentation between all the other keyword argument descriptions.

**Comment by benvanwerkhoven** [comment]
At least the API documentation is visible again! haha
